pinot-controller/src/main/java/org/apache/pinot/controller/api/upload/ZKOperator.java

```diff
@@ -156,7 +167,10 @@ private void processExistingSegment(SegmentMetadata segmentMetadata, URI finalSe
         // (creation time is not included in the crc)
         existingSegmentZKMetadata.setCreationTime(segmentMetadata.getIndexCreationTime());
         existingSegmentZKMetadata.setRefreshTime(System.currentTimeMillis());
-        if (!_pinotHelixResourceManager.updateZkMetadata(tableNameWithType, existingSegmentZKMetadata)) {
+        // NOTE: in rare cases the segment can be deleted before the metadata is updated, we should fail the request for
```

Comment:
Let's just fail the upload when the version does not match regardless of whether the flag is set. This is consistent with the behavior of locking the segment, and more robust. During the locking period, if the metadata changed, just abort the upload.

Reply:
Good point, updated to always check the version.